src/buf/bounded.rs

```diff
@@ -0,0 +1,167 @@
+use super::{IoBuf, IoBufMut, Slice};
+
+use std::ops;
+
+/// A possibly bounded view into an owned [`IoBuf`] buffer.
+///
+/// Because buffers are passed by ownership to the runtime, Rust's slice API
+/// (`&buf[..]`) cannot be used. Instead, `tokio-uring` provides an owned slice
+/// API: [`.slice()`]. The method takes ownership of the buffer and returns a
+/// [`Slice`] value that tracks the requested range.
+///
+/// This trait provides a generic way to use buffers and `Slice` views
+/// into such buffers with `io-uring` operations.
+///
+/// [`.slice()`]: BoundedBuf::slice
+pub trait BoundedBuf: Unpin + 'static {
+    /// The type of the underlying buffer.
+    type Buf: IoBuf;
+
+    /// The type representing the range bounds of the view.
+    type Bounds: ops::RangeBounds<usize>;
+
+    /// Returns a view of the buffer with the specified range.
+    ///
+    /// This method is similar to Rust's slicing (`&buf[..]`), but takes
+    /// ownership of the buffer. The range bounds are specified against
+    /// the possibly offset beginning of the `self` view into the buffer
+    /// and the end bound, if specified, must not exceed the view's total size.
+    /// Note that the range may extend into the uninitialized part of the
+    /// buffer, but it must start (if so bounded) in the initialized part
+    /// or immediately adjacent to it.
+    ///
+    /// # Panics
+    ///
+    /// If the range is invalid with regard to the recipient's total size or
+    /// the length of its initialized part, the implementation of this method
+    /// should panic.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_uring::buf::BoundedBuf;
+    ///
+    /// let buf = b"hello world".to_vec();
+    /// let slice = buf.slice(5..10);
+    /// assert_eq!(&slice[..], b" worl");
+    /// let slice = slice.slice(1..3);
+    /// assert_eq!(&slice[..], b"wo");
+    /// ```
+    fn slice(self, range: impl ops::RangeBounds<usize>) -> Slice<Self::Buf>;
+
+    /// Returns a `Slice` with the view's full range.
+    ///
+    /// This method is to be used by the `tokio-uring` runtime and it is not
+    /// expected for users to call it directly.
+    fn slice_full(self) -> Slice<Self::Buf>;
+
+    /// Gets a reference to the underlying buffer.
+    fn get_buf(&self) -> &Self::Buf;
+
+    /// Returns the range bounds for this view.
+    fn bounds(&self) -> Self::Bounds;
+
+    /// Constructs a view from an underlying buffer and range bounds.
+    fn from_buf_bounds(buf: Self::Buf, bounds: Self::Bounds) -> Self;
+
+    /// Like [`IoBuf::stable_ptr`],
+    /// but possibly offset to the view's starting position.
+    fn stable_ptr(&self) -> *const u8;
+
+    /// Number of initialized bytes available via this view.
+    fn bytes_init(&self) -> usize;
+
+    /// Total size of the view, including uninitialized memory, if any.
+    fn bytes_total(&self) -> usize;
+}
+
+impl<T: IoBuf> BoundedBuf for T {
```

Comment:
Is it possible to do the reverse here, and accepting IoBuf for all of our APIs, with IoBuf being implemented for BoundedBuf implementations?

Reply:
In the currently proposed design, `IoBuf` is an open trait which represents whole buffers, so it excludes `Slice`, while `BoundedBuf` is a sealed extension trait, adds slicing and is also implemented for `Slice`. @Noah-Kennedy are you suggesting to rename these traits, or to make more substantial changes?

Reply:
Ah, that makes sense.

I was thinking that IoBuf would semantically represent anything which can be used as a buffer for an operation (as is currently the case).

Is there a reason that this can't be the case?

Reply:
> I was thinking that IoBuf would semantically represent anything which can be used as a buffer for an operation (as is currently the case).

That role is to be taken by `BoundedBuf`. `IoBuf` is dedicated to representing basic buffers that have stable addresses and length and can maintain the `.bytes_init()` marker, but are otherwise opaque to this crate. `IoBuf` remains an extension point, open for implementation by other crates, while the implementations of `BoundedBuf`  are effectively sealed by tokio-uring.

> Is there a reason that this can't be the case?

We need to be able to pass all types that `slice` can be called on, but for `Slice` this method needs to return `Self`. So we can't have the same trait be an open-ended abstraction for buffers and also generalize the slicing functionality.

Reply:
Ah, makes sense then!